adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -887,7 +887,7 @@ protected void onCommandReceived(final ProtonSender sender, final CommandContext
                             resource.getTenantId(),
                             outcome,
                             context.isRemotelySettled() ? QoS.AT_MOST_ONCE : QoS.AT_LEAST_ONCE,
-                            context.getPayloadSize(),
+                            calculatePayloadSize(context.getPayloadSize(), tenantTracker.result()),
```

Comment:
This means that we *always* charge the client at least the real payload size, right?
If we experience a `ServerErrorException` then I think that we should report a payload size of 0 in order to not charge the client at all, because in this case it is not the client's fault that we cannot process the message, is it?

Reply:
very valid point. I will update it.

Reply:
I guess not for all `ServerErrorException`. The one with cause as `no credit available` due to availability issues of the north bound applications may need to be charged. WDYT?

Reply:
In general I would agree. However, this would require quite some special error handling. Do you have an idea how we would do that efficiently?

Reply:
> This means that we _always_ charge the client at least the real payload size, right?
> If we experience a `ServerErrorException` then I think that we should report a payload size of 0 in order to not charge the client at all, because in this case it is not the client's fault that we cannot process the message, is it?

If we experience a `ServerErrorException`, then we normally record the processing outcome as `UNDELIVERABLE` in the metrics. When we aggregate data from the metrics back-end to verify the message limit, we look for the metrics data with processing outcomes as `FORWARDED` or `UNPROCESSABLE`. So I don't think we need to make the payload as 0 in case of `ServerErrorException` as processing outcome `UNDELIVERABLE` is anyway not part of the message limit calculation.


> In general I would agree. However, this would require quite some special error handling. Do you have an idea how we would do that efficiently?

With this current setup, it seems to me that the only way is to identify based on the error message `no credit available`and I don't find it nice to make use of it. As an alternative, similar to  the `ResourceConflictException` which extends `ClientErrorException`, a new Exception namely `NoCreditAvailableException` can be sub-classed from the `ServerErrorException`. This can be used by the adapters to record the processing outcome as `UNPROCESSABLE` in case of `no credit available`. 

But I am not sure that `no credit available` issue is always due to the availability issues of _northbound applications_ or _devices_. @calohmn Whats your view on this?

Reply:
I think a `no credit available` exception when trying to send a event/telemetry message downstream from a protocol adapter will *probably* mean the northbound application isn't connected. There may be the case however that the error stems from the device registry not giving out credits (maybe due to heavy load) or the AMQP network (e.g. Qpid) not giving out credits for a connected application. Such cases probably point to issues with the Hono setup and scaling of components.

Reply:
Then why don't we create a `NoDownstreamCreditException` instead and only use that in the downstream senders when they have no credit? We could stick to the existing pattern used in the device registry clients where we create a `ServerErrorException` with code 503 instead.

Reply:
> Then why don't we create a `NoDownstreamCreditException` instead and only use that in the downstream senders when they have no credit? We could stick to the existing pattern used in the device registry clients where we create a `ServerErrorException` with code 503 instead.

I would suggest a separate PR for implementing the `NoDownstreamCreditException` and recording the metrics with outcome as `UNPROCESSABLE` in case of this scenario.

I would continue with this PR by moving the business message payload size calculation into a central location (inside `reportTelemetry(...)` and `reportCommand(...)` methods inside `Metrics`).